wiring/src/spark_wiring_logging.cpp

```diff
@@ -0,0 +1,246 @@
+            if (handlers_.size() == 1) {
+                log_set_callbacks(nullptr, nullptr, nullptr, nullptr);
+            }
+            handlers_.erase(it);
+        }
+    }
+
+    static LogManager* instance() {
+        static LogManager mgr;
+        return &mgr;
+    }
+
+private:
+    std::vector<LogHandler*> handlers_;
+
+    static void logMessage(const char *msg, int level, const char *category, uint32_t time, const char *file, int line,
+            const char *func, void *reserved) {
+        const LogHandler::SourceInfo info = {file, func, line};
+        const auto &handlers = instance()->handlers_;
+        for (size_t i = 0; i < handlers.size(); ++i) {
+            handlers[i]->message(msg, (LogLevel)level, category, time, info);
+        }
+    }
+
+    static void logWrite(const char *data, size_t size, int level, const char *category, void *reserved) {
+        const auto &handlers = instance()->handlers_;
+        for (size_t i = 0; i < handlers.size(); ++i) {
+            handlers[i]->write(data, size, (LogLevel)level, category);
+        }
+    }
+
+    static int logEnabled(int level, const char *category, void *reserved) {
+        int minLevel = LOG_LEVEL_NONE;
+        const auto &handlers = instance()->handlers_;
+        for (size_t i = 0; i < handlers.size(); ++i) {
+            const int level = handlers[i]->categoryLevel(category);
+            if (level < minLevel) {
+                minLevel = level;
+            }
+        }
+        return (level >= minLevel);
+    }
+};
+
+inline const char* strchrend(const char* s, char c) {
+    const char* result = strchr(s, c);
+    return result ? result : s + strlen(s);
+}
+
+} // namespace
+
+// spark::LogHandler
+struct spark::LogHandler::FilterData {
+    const char *name; // Category name
+    size_t size; // Name length
+    int level; // Logging level (-1 if not specified for this category)
+    std::vector<FilterData> filters; // Subcategories
+
+    FilterData(const char *name, size_t size) :
+            name(name),
+            size(size),
+            level(-1) {
+    }
+};
+
+spark::LogHandler::LogHandler(Stream *stream, LogLevel level, const Filters &filters) :
+        stream_(stream),
+        level_(level) {
+    for (auto it = filters.begin(); it != filters.end(); ++it) {
+        const char* const category = it->first;
+        const LogLevel level = it->second;
+        std::vector<FilterData> *filters = &filters_; // Root categories
+        size_t pos = 0;
+        for (size_t i = 0;; ++i) {
+            if (category[i] && category[i] != '.') { // Category separator
+                continue;
+            }
+            const size_t size = i - pos;
+            if (!size) {
+                break; // Invalid category name
+            }
+            const char* const name = category + pos;
+            bool found = false;
+            auto it = std::lower_bound(filters->begin(), filters->end(), std::make_pair(name, size),
```

Comment:
This seems like a good candidate for a method extraction: `bool doStuff(const Filters *filters, const std::pair<const char*, size_t> &value) {...}` where `doStuff` is an appropriate name for the stuff this block does.